File: bdp-devops/ambari/ambari-web/app/models/configs/theme/sub_section.js
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

var App = require('app');

App.SubSection = DS.Model.extend({

  id: DS.attr('string'),

  /**
   * @type {string}
   */
  name: DS.attr('string'),

  /**
   * theme from which this is coming from , eg: default, database, credentials, etc.
   */
  themeName: DS.attr('string'),

  /**
   * @type {string}
   */
  displayName: DS.attr('string'),

  /**
   * @type {boolean}
   */
  border: DS.attr('boolean', {defaultValue: false}),

  /**
   * @type {number}
   */
  rowIndex: DS.attr('number', {defaultValue: 1}),

  /**
   * @type {number}
   */
  columnIndex: DS.attr('number', {defaultValue: 1}),

  /**
   * @type {number}
   */
  rowSpan: DS.attr('number', {defaultValue: 1}),

  /**
   * @type {number}
   */
  columnSpan: DS.attr('number', {defaultValue: 1}),

  /**
   * @type {App.Section}
   */
  section: DS.belongsTo('App.Section'),

  /**
   * @type {String[]}
   */
  configProperties: DS.attr('array', {defaultValue: []}),

  /**
   * @type {App.SubSectionTab[]}
   */
  subSectionTabs: DS.hasMany('App.SubSectionTab'),


  dependsOn: DS.attr('array', {defaultValue: []}),

  /**
   * @type {boolean}
   */
  leftVerticalSplitter: DS.attr('boolean', {defaultValue: true}),

  /**
   * @type {App.ServiceConfigProperty[]}
   */
  configs: [],

  /**
   * @type {boolean}
   */
  hasTabs: Em.computed.bool('subSectionTabs.length'),

  someSubSectionTabIsVisible: Em.computed.someBy('subSectionTabs', 'isVisible', true),

  showTabs: Em.computed.and('hasTabs', 'someSubSectionTabIsVisible'),

  visibleProperties: function() {
    return this.get('configs').filter(function(c) {
      return c.get('isVisible') && !c.get('hiddenBySection');
    });
  }.property('configs.@each.isVisible', 'configs.@each.hiddenBySection'),

  visibleTabs: Em.computed.filterBy('subSectionTabs', 'isVisible', true),

  /**
   * Number of the errors in all configs
   * @type {number}
   */
  errorsCount: function () {
    var propertiesWithErrors = this.get('visibleProperties').filter(function(c) {
      return !c.get('isValid') || !c.get('isValidOverride');
    }).length;
    var tabsWithErrors = this.get('visibleTabs').mapProperty('errorsCount').reduce(Em.sum, 0);
    return propertiesWithErrors + tabsWithErrors;
  }.property('visibleProperties.@each.isValid', 'visibleProperties.@each.isValidOverride', 'visibleTabs.@each.errorsCount'),

  /**
   * @type {boolean}
   */
  addLeftVerticalSplitter: Em.computed.and('!isFirstColumn', 'leftVerticalSplitter'),

  /**
   * @type {boolean}
   */
  addRightVerticalSplitter: Em.computed.not('isLastColumn'),

  /**
   * @type {boolean}
   */
  showTopSplitter: Em.computed.and('!isFirstRow', '!border'),

  /**
   * @type {boolean}
   */
  isFirstRow: Em.computed.equal('rowIndex', 0),

  /**
   * @type {boolean}
   */
  isMiddleRow: function () {
    return this.get('rowIndex') != 0 && (this.get('rowIndex') + this.get('rowSpan') < this.get('section.sectionRows'));
  }.property('rowIndex', 'rowSpan', 'section.sectionRows'),

  /**
   * @type {boolean}
   */
  isLastRow: function () {
    return this.get('rowIndex') + this.get('rowSpan') == this.get('section.sectionRows');
  }.property('rowIndex', 'rowSpan', 'section.sectionRows'),

  /**
   * @type {boolean}
   */
  isFirstColumn: Em.computed.equal('columnIndex', 0),

  /**
   * @type {boolean}
   */
  isMiddleColumn: function () {
    return this.get('columnIndex') != 0 && (this.get('columnIndex') + this.get('columnSpan') < this.get('section.sectionColumns'));
  }.property('columnIndex', 'columnSpan', 'section.sectionColumns'),

  /**
   * @type {boolean}
   */
  isLastColumn: function () {
    return this.get('columnIndex') + this.get('columnSpan') == this.get('section.sectionColumns');
  }.property('columnIndex', 'columnSpan', 'section.sectionColumns'),

  /**
   * If the visibility of subsection is dependent on a value of some config
   */
  isHiddenByConfig: false,

  /**
   * Determines if subsection is filtered by checking it own configs
   * If there is no configs, subsection can't be hidden
   * @type {boolean}
   */
  isHiddenByFilter: function () {
    var configs = this.get('configs').filter(function(c) {
      return !c.get('hiddenBySection') && c.get('isVisible');
    });
    return configs.length ? configs.everyProperty('isHiddenByFilter', true) && !this.get('someSubSectionTabIsVisible'): false;
  }.property('configs.@each.isHiddenByFilter'),

  /**
   * @type {boolean}
   */
  someConfigIsVisible: Em.computed.someBy('configs', 'isVisible', true),

  /**
   * Determines if subsection is visible
   * @type {boolean}
   */
  isSectionVisible: Em.computed.and('!isHiddenByFilter', '!isHiddenByConfig', 'someConfigIsVisible')

});


App.SubSection.FIXTURES = [];

File: bdp-devops/ambari/ambari-server/src/main/java/org/apache/ambari/server/state/theme/Subsection.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.ambari.server.state.theme;

import java.util.List;

import org.codehaus.jackson.annotate.JsonIgnoreProperties;
import org.codehaus.jackson.annotate.JsonProperty;
import org.codehaus.jackson.map.annotate.JsonSerialize;

import io.swagger.annotations.ApiModelProperty;

@JsonSerialize(include= JsonSerialize.Inclusion.NON_NULL)
@JsonIgnoreProperties(ignoreUnknown = true)
public class Subsection {
	@JsonProperty("row-index")
	private String rowIndex;
	@JsonProperty("name")
	private String name;
  @JsonProperty("display-name")
	private String displayName;
	@JsonProperty("column-span")
	private String columnSpan;
	@JsonProperty("row-span")
	private String rowSpan;
	@JsonProperty("column-index")
	private String columnIndex;
  @JsonProperty("border")
	private String border;
  @JsonProperty("left-vertical-splitter")
  private Boolean leftVerticalSplitter;
  @JsonProperty("depends-on")
  private List<ConfigCondition> dependsOn;
  @JsonProperty("subsection-tabs")
  private List<SubsectionTab> subsectionTabs;


  @ApiModelProperty( name = "row-index")
  public String getRowIndex() {
    return rowIndex;
  }

  public void setRowIndex(String rowIndex) {
    this.rowIndex = rowIndex;
  }

  @ApiModelProperty( name = "name")
  public String getName() {
    return name;
  }

  public void setName(String name) {
    this.name = name;
  }

  @ApiModelProperty( name = "column-span")
  public String getColumnSpan() {
    return columnSpan;
  }

  public void setColumnSpan(String columnSpan) {
    this.columnSpan = columnSpan;
  }

  @ApiModelProperty( name = "row-span")
  public String getRowSpan() {
    return rowSpan;
  }

  public void setRowSpan(String rowSpan) {
    this.rowSpan = rowSpan;
  }

  @ApiModelProperty( name = "column-index")
  public String getColumnIndex() {
    return columnIndex;
  }

  public void setColumnIndex(String columnIndex) {
    this.columnIndex = columnIndex;
  }

  @ApiModelProperty( name = "display-name")
  public String getDisplayName() {
    return displayName;
  }

  public void setDisplayName(String displayName) {
    this.displayName = displayName;
  }

  @ApiModelProperty( name = "border")
  public String getBorder() {
    return border;
  }

  public void setBorder(String border) {
    this.border = border;
  }

  @ApiModelProperty( name = "left-vertical-splitter")
  public Boolean getLeftVerticalSplitter() {
    return leftVerticalSplitter;
  }

  public void setLeftVerticalSplitter(Boolean leftVerticalSplitter) {
    this.leftVerticalSplitter = leftVerticalSplitter;
  }

  @ApiModelProperty( name = "depends-on")
  public List<ConfigCondition> getDependsOn() {
    return dependsOn;
  }

  public void setDependsOn(List<ConfigCondition> dependsOn) {
    this.dependsOn = dependsOn;
  }

  @ApiModelProperty( name = "subsection-tab")
  public List<SubsectionTab> getSubsectionTabs() {
    return subsectionTabs;
  }

  public void setSubsectionTabs(List<SubsectionTab> subsectionTabs) {
    this.subsectionTabs = subsectionTabs;
  }

  @ApiModelProperty( name = "removed")
  public boolean isRemoved() {
    return rowIndex == null && rowSpan == null && columnIndex == null && columnSpan == null && dependsOn == null && subsectionTabs == null;
  }

  public void mergeWithParent(Subsection parent) {
    if (rowSpan == null) {
      rowSpan = parent.rowSpan;
    }
    if (rowIndex == null) {
      rowIndex = parent.rowIndex;
    }
    if (columnSpan == null) {
      columnSpan = parent.columnSpan;
    }
    if (columnIndex == null) {
      columnIndex = parent.columnIndex;
    }
    if (displayName == null) {
      displayName = parent.displayName;
    }
    if (border == null) {
      border = parent.border;
    }
    if (leftVerticalSplitter == null) {
      leftVerticalSplitter = parent.leftVerticalSplitter;
    }
    if (dependsOn == null) {
      dependsOn = parent.dependsOn;
    }
    if (subsectionTabs == null) {
      subsectionTabs = parent.subsectionTabs;
    }
  }

  @JsonSerialize(include = JsonSerialize.Inclusion.NON_NULL)
  @JsonIgnoreProperties(ignoreUnknown = true)
  private static class SubsectionTab {
    @JsonProperty("name")
    private String name;
    @JsonProperty("display-name")
    private String displayName;
    @JsonProperty("depends-on")
    private List<ConfigCondition> dependsOn;

    public String getName() {
      return name;
    }

    public void setName(String name) {
      this.name = name;
    }

    public String getDisplayName() {
      return displayName;
    }

    public void setDisplayName(String displayName) {
      this.displayName = displayName;
    }

    public List<ConfigCondition> getDependsOn() {
      return dependsOn;
    }

    public void setDependsOn(List<ConfigCondition> dependsOn) {
      this.dependsOn = dependsOn;
    }

  }
}